### PR TITLE
Adding long decimal support for arbitrary() function

### PR DESCRIPTION
**Pull request by minhancao**
Delivers https://github.com/facebookincubator/velox/issues/7311

### CODE REVIEW

**Review by aditi-pandit — APPROVED** [review]
Thanks @minhancao

**Review by aditi-pandit — CHANGES_REQUESTED** [review]
@minhancao : Please can you run AggregationFuzzer and post results from it as well https://facebookincubator.github.io/velox/develop/testing/fuzzer.html#aggregation-fuzzer

**Comment by minhancao** [comment]
@aditi-pandit 
Here is the result for the aggregation fuzzer test for arbitrary():


I ran `_build/debug/velox/functions/prestosql/fuzzer/velox_aggregation_fuzzer_test --seed 123456 --duration_sec 60 --logtostderr=1 --minloglevel=0 --only arbitrary --v=1 2>&1 | tee agg_fuzzer_test_arbitrary.out`

I can provide the whole output file if needed.

```
I20240327 10:43:07.766820 4177649 AggregationFuzzer.cpp:453] ==============================> Done with iteration 22
I20240327 10:43:07.766849 4177649 AggregationFuzzer.cpp:992] Total masked aggregations: 3 (13.04%)
I20240327 10:43:07.766871 4177649 AggregationFuzzer.cpp:994] Total global aggregations: 0 (0.00%)
I20240327 10:43:07.766885 4177649 AggregationFuzzer.cpp:996] Total group-by aggregations: 20 (86.96%)
I20240327 10:43:07.766901 4177649 AggregationFuzzer.cpp:998] Total distinct aggregations: 3 (13.04%)
I20240327 10:43:07.766918 4177649 AggregationFuzzer.cpp:1000] Total aggregations over distinct inputs: 3 (13.04%)
I20240327 10:43:07.766933 4177649 AggregationFuzzer.cpp:1002] Total window expressions: 0 (0.00%)
I20240327 10:43:07.766947 4177649 AggregationFuzzerBase.cpp:557] Total functions tested: 1
I20240327 10:43:07.766960 4177649 AggregationFuzzerBase.cpp:558] Total iterations requiring sorted inputs: 3 (13.04%)
I20240327 10:43:07.766974 4177649 AggregationFuzzerBase.cpp:560] Total iterations verified against reference DB: 4 (17.39%)
I20240327 10:43:07.766988 4177649 AggregationFuzzerBase.cpp:562] Total functions not verified (verification skipped / not supported by reference DB / reference DB failed): 17 (73.91%) / 2 (8.70%) / 0 (0.00%)
I20240327 10:43:07.767002 4177649 AggregationFuzzerBase.cpp:567] Total failed functions: 0 (0.00%)
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

**Review by aditi-pandit — APPROVED** [review]
Thanks @minhancao

**Comment by aditi-pandit** [comment]
@Yuhta @kagamiori : Please can you help with review and merge.

File: velox/functions/prestosql/aggregates/ArbitraryAggregate.cpp [hunk 1]
```diff
@@ -294,6 +294,11 @@ void registerArbitraryAggregate(
             return std::make_unique<ArbitraryAggregate<int32_t>>(inputType);
           case TypeKind::BIGINT:
             return std::make_unique<ArbitraryAggregate<int64_t>>(inputType);
+          case TypeKind::HUGEINT:
+            if (inputType->isLongDecimal()) {
+              return std::make_unique<ArbitraryAggregate<int128_t>>(inputType);
+            }
+            VELOX_UNREACHABLE();
```
Comment (karteekmurthys) [inline velox/functions/prestosql/aggregates/ArbitraryAggregate.cpp:301]: Replace with `VELOX_NYI`

File: velox/functions/prestosql/aggregates/tests/ArbitraryTest.cpp [hunk 2]
```diff
@@ -349,6 +349,60 @@ TEST_F(ArbitraryTest, interval) {
...
+                                  std::nullopt,
+                                  HugeInt::build(10, 400)},
+                                 DECIMAL(38, 8))});
+
+  auto expectedResult = makeRowVector({
+      makeFlatVector<int64_t>({1, 2, 3, 4}),
+      makeNullableFlatVector<int128_t>(
+          {HugeInt::build(10, 100),
+           HugeInt::build(10, 200),
+           std::nullopt,
```
Comment (karteekmurthys) [inline velox/functions/prestosql/aggregates/tests/ArbitraryTest.cpp:371]: From presto doc:
```
arbitrary(x) → [same as input][#](https://prestodb.io/docs/current/functions/aggregate.html#arbitrary)
Returns an arbitrary non-null value of x, if one exists.
```
Would you please verify this behavior is consistent with Presto JAVA.

File: velox/functions/prestosql/aggregates/ArbitraryAggregate.cpp (hunk 1 above)
Comment (minhancao) [reply velox/functions/prestosql/aggregates/ArbitraryAggregate.cpp]: Done

File: velox/functions/prestosql/aggregates/tests/ArbitraryTest.cpp (hunk 2 above)
Comment (minhancao) [reply velox/functions/prestosql/aggregates/tests/ArbitraryTest.cpp:371]: The behavior is indeed consistent with Presto JAVA.

Presto java:
```
presto:sf1> SELECT arbitrary(Col) FROM (VALUES cast(0.82 as decimal(5,4)), cast(2.333 as decimal(5,4)), cast(3.132 as decimal(5,4)), cast(4.344 as decimal(38,4))) AS X(Col);
 _col0  
--------
 0.8200 
(1 row)

presto:sf1> select arbitrary(cast(3.3 as decimal(38,4)));
 _col0  
--------
 3.3000 
(1 row)
```

Velox c++:
```
presto:tpch> SELECT arbitrary(Col) FROM (VALUES cast(0.82 as decimal(5,4)), cast(2.333 as decimal(5,4)), cast(3.132 as decimal(5,4)), cast(4.344 as decimal(38,4))) AS X(Col);
 _col0  
--------
 0.8200 
(1 row)

presto:tpch> select arbitrary(cast(3.3 as decimal(38,4)));
 _col0  
--------
 3.3000 
(1 row)
```

Comment (karteekmurthys) [reply velox/functions/prestosql/aggregates/tests/ArbitraryTest.cpp:371]: Sorry, I should have been more specific. Validate this test case in JAVA:

```
  makeFlatVector<int64_t>({1, 1, 2, 2, 3, 3, 4, 4}),
                             makeNullableFlatVector<int128_t>(
                                 {HugeInt::build(10, 100),
                                  HugeInt::build(10, 100),
                                  HugeInt::build(10, 200),
                                  HugeInt::build(10, 200),
                                  std::nullopt,
                                  std::nullopt,
                                  std::nullopt,
                                  HugeInt::build(10, 400)},
                                 DECIMAL(38, 8))});
```

Comment (minhancao) [reply velox/functions/prestosql/aggregates/tests/ArbitraryTest.cpp:371]: Validated the testcase, both Presto Java and Velox C++ are behaving the same:

Presto Java:
```
presto> SELECT arbitrary(value), key 
     -> FROM 
     -> ( VALUES 
     -> (1, cast(cast(100 as BIGINT) as decimal(38,8))), (1, cast(cast(100 as BIGINT) as decimal(38,8))), (2, cast(cast(200 as BIGINT) as decimal(38,8))), (2, cast(cast(200 as BIGINT) as decimal(38,8))), (3, null), (3, null), (4, cast(cast(400 as BIGINT) as decimal(38,8))), (4, cast(cast(400 as BIGINT) as decimal(38,8))) ) 
     -> AS 
     -> t(key, value) group by key;
    _col0     | key 
--------------+-----
 NULL         |   3 
 200.00000000 |   2 
 400.00000000 |   4 
 100.00000000 |   1 
(4 rows)
```

Velox C++:
```
presto:tpch> SELECT arbitrary(value), key 
          -> FROM 
          -> ( VALUES 
          -> (1, cast(cast(100 as BIGINT) as decimal(38,8))), (1, cast(cast(100 as BIGINT) as decimal(38,8))), (2, cast(cast(200 as BIGINT) as decimal(38,8))), (2, cast(cast(200 as BIGINT) as decimal(38,8))), (3, null), (3, null), (4, cast(cast(400 as BIGINT) as decimal(38,8))), (4, cast(cast(400 as BIGINT) as decimal(38,8))) ) 
          -> AS 
          -> t(key, value) group by key;
    _col0     | key 
--------------+-----
 400.00000000 |   4 
 100.00000000 |   1 
 NULL         |   3 
 200.00000000 |   2 
(4 rows)
```

**Review by karteekmurthys — APPROVED** [review]
Thanks @minhancao!

**Review by karteekmurthys — COMMENTED** [review]
[ RUN      ] ArbitraryTest.longDecimal
*** Aborted at 1712092248 (Unix time, try 'date -d @1712092248') ***
*** Signal 11 (SIGSEGV) (0x0) received by PID 19932 (pthread TID 0x7f87f4bff700) (linux TID 57052) (code: 128), stack trace: ***
(error retrieving stack trace)

test 294
        Start 294: velox_s3registration_test

Test failure. Likely we may have to switchto aligned memory allocator.

**Comment by minhancao** [comment]
@karteekmurthys Thank you for identifying the test failure above! I was wondering how the switch to aligned memory allocator should be approached or implemented?

Would it be similar to what you have in your PR from here? https://github.com/facebookincubator/velox/pull/8723/files#diff-5091b1dd8232c357067be47fb0e751772f91a0aa12dea5db835a2d4ae296a63cR93

**Comment by karteekmurthys** [comment]
> @karteekmurthys Thank you for identifying the test failure above! I was wondering how the switch to aligned memory allocator should be approached or implemented?
> 
> Would it be similar to what you have in your PR from here? https://github.com/facebookincubator/velox/pull/8723/files#diff-5091b1dd8232c357067be47fb0e751772f91a0aa12dea5db835a2d4ae296a63cR93

Yes. likely. Please try it.

**Comment by karteekmurthys** [comment]
@minhancao would you please try overriding this method:

```
 int32_t accumulatorAlignmentSize() const override {
    return static_cast<int32_t>(sizeof(int128_t));
  }

```

**Comment by minhancao** [comment]
@karteekmurthys I have implemented the change above, how do I get the linux CircleCI test to run on this PR?

File: velox/functions/prestosql/aggregates/ArbitraryAggregate.cpp [hunk 3]
```diff
@@ -39,6 +39,10 @@ class ArbitraryAggregate : public SimpleNumericAggregate<T, T, T> {
     return sizeof(T);
   }
 
+  int32_t accumulatorAlignmentSize() const override {
```
Comment (karteekmurthys) [inline velox/functions/prestosql/aggregates/ArbitraryAggregate.cpp:42]: This needs to return the size of 128bit. And you need only to override it for ArbitraryAggregate<> where the types are INT128.

**Comment by karteekmurthys** [comment]
@kagamiori would you please review this.

**Comment by karteekmurthys** [comment]
@Yuhta would you please help with the review.

**Comment by minhancao** [comment]
Hi @kagamiori @Yuhta , can you review this PR when you can please?

**Review by kagamiori — APPROVED** [review]
LGTM. Thank you for adding this support!

**Pull request closed without merging**